lib/topological_inventory/orchestrator/metric_scaler.rb

```diff
@@ -0,0 +1,88 @@
+require "manageiq-loggers"
+
+module TopologicalInventory
+  module Orchestrator
+    class MetricScaler
+      attr_reader :logger
+
+      def initialize(logger = nil)
+        @logger = logger || ManageIQ::Loggers::CloudWatch.new
+      end
+
+      def run
+        loop do
+          run_once
+
+          sleep 10
+        end
+      end
+
+      def run_once
+        logger.info("#{self.class.name}##{__method__} Starting...")
+        object_manager.get_deployment_configs("metric_scaler_enabled=true").each do |dc|
+          current_metric_name = dc.metadata.labels["metric_scaler_current_metric_name"]       # i.e. "topological_inventory_api_puma_busy_threads"
+          max_metric_name     = dc.metadata.labels["metric_scaler_max_metric_name"]           # i.e. "topological_inventory_api_puma_max_threads"
+          max_replicas        = dc.metadata.labels["metric_scaler_max_replicas"]&.to_i        # i.e. 5
+          min_replicas        = dc.metadata.labels["metric_scaler_min_replicas"]&.to_i        # i.e. 1
+          target_usage_pct    = dc.metadata.labels["metric_scaler_target_usage_pct"]&.to_i    # i.e. 50
+          scale_threshold_pct = dc.metadata.labels["metric_scaler_scale_threshold_pct"]&.to_i # i.e. 20
+
+          next unless current_metric_name && max_metric_name && max_replicas && min_replicas && target_usage_pct && scale_threshold_pct
+          logger.info("Metrics scaling enabled for #{dc.metadata.name}")
+
+          endpoint = object_manager.get_endpoint(dc.metadata.name)
+          pod_ips  = endpoint.subsets.flat_map { |s| s.addresses.collect { |a| a[:ip] } }
+
+          total_consumed = 0
+          total_max      = 0
+
+          pod_ips.each do |ip|
+            h = scrape_metrics_from_ip(ip)
+            total_consumed += h[current_metric_name].to_f
+            total_max      += h[max_metric_name].to_f
+          end
+
+          current_usage_pct = (total_consumed.to_f / total_max.to_f) * 100
+          deviation_pct = current_usage_pct - target_usage_pct
+
+          logger.info("#{dc.metadata.name} consuming #{total_consumed} of #{total_max}, #{current_usage_pct}%")
+
+          next if deviation_pct < scale_threshold_pct # Within tolerance
```

Comment:
I think this should be an absolute value of the deviation_pct as it can be negative. Otherwise the tolerance wouldn't apply when we scale down.